compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -412,20 +346,603 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            // We'll try to detect 4 different types of mistakes:
+            // - An extra parameter has been provided that doesn't satisfy *any* of the other inputs
+            // - An input is missing, which isn't satisfied by *any* of the other arguments
+            // - Some number of arguments have been provided in the wrong order
+            // - A type is straight up invalid
+
+            // First, fill in the rest of our compatibility matrix
+            for i in 0..provided_arg_count {
+                for j in 0..minimum_input_count {
+                    if i == j {
+                        continue;
+                    }
+                    compatibility_matrix[i][j] = check_compatible(i, j);
+                }
+            }
+
+            // Next, let's construct the error
+            let (error_span, ctor_of) = match &call_expr.kind {
+                hir::ExprKind::Call(
+                    hir::Expr {
+                        span,
+                        kind:
+                            hir::ExprKind::Path(hir::QPath::Resolved(
+                                _,
+                                hir::Path { res: Res::Def(DefKind::Ctor(of, _), _), .. },
+                            )),
+                        ..
+                    },
+                    _,
+                ) => (*span, Some(of)),
+                hir::ExprKind::Call(hir::Expr { span, .. }, _) => (*span, None),
+                hir::ExprKind::MethodCall(_, span, _, _) => (
+                    *span, None, // methods are never ctors
+                ),
+                k => span_bug!(call_span, "checking argument types on a non-call: `{:?}`", k),
+            };
+            let call_name = match ctor_of {
+                Some(CtorOf::Struct) => "struct",
+                Some(CtorOf::Variant) => "enum variant",
+                None => "function",
+            };
+            if c_variadic && provided_arg_count < minimum_input_count {
+                err_code = "E0060";
+            }
+            let mut err = if minimum_input_count == provided_arg_count {
+                struct_span_err!(
+                    tcx.sess,
+                    error_span,
+                    E0308,
+                    "arguments to this {} are incorrect",
+                    call_name,
+                )
+            } else {
+                tcx.sess.struct_span_err_with_code(
+                    error_span,
+                    &format!(
+                        "this {} takes {}{} but {} {} supplied",
+                        call_name,
+                        if c_variadic { "at least " } else { "" },
+                        potentially_plural_count(minimum_input_count, "argument"),
+                        potentially_plural_count(provided_arg_count, "argument"),
+                        if provided_arg_count == 1 { "was" } else { "were" }
+                    ),
+                    DiagnosticId::Error(err_code.to_owned()),
+                )
```

Comment:
This could be its own change, as it isn't reliant on the subsequent checks.

Reply:
In #93118, I make error emission a bit more lazy and more like in this PR. It's not a 100% transplant, but close enough the diff should be nicer.